### DISCUSSION TITLE
SSG and .next folder with a load balancer and multiple Nextjs instances

### DISCUSSION THREAD

**Discussion by ExtremelyAngryCat** [vercel/next.js#19773] [Help] [answered]
Hi,

I have an environment with muliple Nextjs containers being handled with a load balancer (via AWS Fargate).
I implemented the iSSG (getStaticProps, getStaticPaths) and it works perfectly with a single Nextjs instance, but with multiples you need a random number of request on a page to trigger the revalidation of every instances (minimum 2 per instance if you are lucky, 1 to trigger the revalidation and the second to actually see the new version).

I noticed the .next/server/pages folder contains a list of files/folders for each generated page, and refreshed with new content after each revalidation. I went looking for more info and i saw this comment: https://github.com/vercel/next.js/discussions/11552#discussioncomment-2804
So i shared a volume (EFS) for all my NextJS instances as the .next folder, but without any success. The folder is used by all instances, i can see the .next/server/pages being written/edited by every instances, but even with a fresh version of the page in the folder, when the other instance hits it still displays the old page version (then write on the .next folder, thus overwritting the file that already is the new version).

So it seems that the .next/server/pages is not being used at all for reading, feeling like a log folder...

**After all this context, my questions are:**

1. Does NextJs have an in-memory per-instance cache for the SSG'd pages, and if so, what is the point of the .next/server/pages folder ?
2. How to handle SSG with multiple instances to not have a mix of old/new page version when reloading a page (i want to handle this first before using robust cdn caching)

Thanks for your time !

**Comment by santiagoesteva** [comment]
+1

**Reply by rrjanbiah** [reply]
@ExtremelyAngryCat Have you found any solution?

**Reply by ExtremelyAngryCat** [reply]
No unfortunately...

We had to do a quite dirty thing to bypass this behavior: 

**Context**: Our nextjs servers are deployed via AWS ECS's, and we actually **know** when there is new versions of the website.

So we implemented a AWS Lamda that list every nextjs tasks to get their private IP's. And when a new version of our site is available and we want all instances to have it, we call that Lambda to call each instance individually, thus triggering the cache regeneration for each of them. But before calling the Lambda, to be sure each instances will revalidate, we must wait X secondes before calling, X being the [revalidate](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration) value. And it works.

As i went trough the doc to get the revalidate infos, i see that they might have implemented new mechanics to handle this properly with On-demand Revalidation and the Self-hosting ISR. The latter might be exactly what you need @MicahRamirez 

I dont know when i will try it, let say the first one who does comes back here to give a feedback :)

**Reply by MicahRamirez** [reply]
> So we implemented a AWS Lamda that list every nextjs tasks to get their private IP's. And when a new version of our site is available and we want all instances to have it, we call that Lambda to call each instance individually, thus triggering the cache regeneration for each of them. But before calling the Lambda, to be sure each instances will revalidate, we must wait X secondes before calling, X being the [revalidate](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration) value. And it works.

Interesting. I imagine the revalidate value is relatively low in your case or waiting that period of time is not necessarily critical in your deployments? 

I had tried the self hosting section, but I had been leveraging their with docker example [here](https://github.com/vercel/next.js/tree/canary/examples/with-docker). 

I'm still seeing the behavior I mentioned below. I had also confirmed that the setting 
```
module.exports = {
  experimental: {
    isrMemoryCacheSize: 0,
  },
}
```
was being passed to the standalone server in the build folder. 

There's a few variables I probably need to try to control for. 
- trying to remove `outputStandalone`
- verify nothing is off on my efs mount

@ExtremelyAngryCat How did you end up handling the mounting the efs mount ? 

For me I had to do the following steps (Using ECS Fargate 1.4)
- Build time copy the .next folder into a tmp directory (`/app/tmp/.next`)
- At runtime, mount NFS on `/app/.next` (bind mount)
- Using an entrypoint script, copy over the build into `/app/.next` (efs mount) if the build_id DNE or mismatches the existing one

In your case was this a bit simpler? Anyways thanks for the update and tag!

**Reply by ExtremelyAngryCat** [reply]
Oh i didnt understand you tried it already, when i created this discussion, the feature didnt exist yet.

> I imagine the revalidate value is relatively low
Yes, a few seconds only

> How did you end up handling the mounting the efs mount ?
I dont remember in detail since its been 1.5 year :'< But i think it was simpler than that, just some comfiguration in the ecs-task-definition to mount a volume, and the sharing between task was some AWS thing too but nothing to do at runtime. Everything was handled on deploy, just had to copy the latest built .next into the shared volume

**Comment by MicahRamirez** [comment]
> So i shared a volume (EFS) for all my NextJS instances as the .next folder, but without any success. The folder is used by all instances, i can see the .next/server/pages being written/edited by every instances, but even with a fresh version of the page in the folder, when the other instance hits it still displays the old page version (then write on the .next folder, thus overwritting the file that already is the new version).

I'm seeing basically the same behavior. 

I have isrMemoryCacheSize: 0. So should be using the FS. I am using EFS.  

Expected Behavior
Fresh build
ALB routes client to instance A client sees `page built at 10:01`
ALB routes client to instance B client sees `page built at 10:01` 

Regardless of the instance routed to, I'd expect the same page built at time visible to the end user. 

Actual Behavior 
Fresh build
ALB routes client to instance A client sees `page built at 10:00`
ALB routes client to instance B client sees `page built at 10:01`

From then on all traffic sees `page built at 10:01`

This feels way more difficult to get working than I expected and the docs are extremely light on this topic.

**Reply by MartinDavi** [reply]
Alright, it's 2023. Any updates now that you've (hopefully) managed to solve this? We're going down the same road right now and need to know what we're in for, haha.

**Comment by umashankar-pardhi-amla** [comment]
Any update on this ?

**Reply by CaptainN** [reply]
From what I understand, the static build cache and ISR caches are separate. But maybe I got that all wrong.

You can [configure ISR cache](https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath), at least in app router (not sure about pages router), and have it store somewhere outside of instance memory, like in redis. But using a shared filesystem store is [specifically supported](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#self-hosting-isr).

If you have multiple containers that are each built separately (instead of just run separately after a singular build), that can cause issues due to mismatched buildId - you can normalize that with [a setting in next.config.js](https://nextjs.org/docs/app/api-reference/next-config-js/generateBuildId):

```js
  // next.config.js
  // Uses a git hash passed in from build env
  generateBuildId: async () => {
    return process.env.GIT_HASH;
  },
```

**Comment by leerob** [comment, marked as answer]
Hey everyone – we have just published new documentation on self-hosting, including how to configure a custom cache handler for use with ISR / Data Cache in both the Pages and App Router. These docs go in depth about how other features like Image Optimization, Middleware, Environment Variables, Build Caching, and more work when self-hosted. There's also updated Docker based examples.

https://nextjs.org/docs/app/building-your-application/deploying
